rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+
+#### Worker
+
+The physical machine(s) containing the physical devices (e.g. GPUs, TPUs) on
+which the replicated computation is executed. A worker may contain one or more
+replicas. Typically one worker will correspond to one host machine, but in the
+case of very large models with model parallelism, one worker may span multiple
+hosts.
+
+#### Sync vs Async replicas
+
+In synchronous training, all replicas are in sync with each other and gradients
+are aggregated across all the replicas. In fully asynchronous training, all
+replicas operate asynchronously and update the variables independently. We can
+also support a hybrid approach, where the replicas are divided into "sync
+groups" - within each sync group, each replica is in sync, but the sync groups
+are asynchronous among themselves. To support this, we later define
+"num_replicas" as the total number of replicas (sync or async) and
+"num_replicas_in_sync" as the number of replicas in one sync group.
+
+#### Input Pipeline
+
+An instance of the input pipeline (e.g. `Dataset`, `FIFOQueue`) obtained by
+calling the user's `input_fn` once. One input pipeline may be used to feed one
+or more model replicas (on one or more workers). See
+[InputReplicationMode](#inputreplicationmode).
+
+### Replicator API
+
+#### Detailed Design
+
+Here are the main ingredients of the API:
+
+*   `Replicator` class: This is the main class of the API and provides the main
+    methods to replicate the computation.
+    *   `__init__:`Creates a Replicator instance. It takes a
+        `DistributionStrategy` object which is used for (almost) all the
+        underlying operations such as replicating the computation, all-reduce
+        etc.
+    *   `scope`: A context manager that manages variable creation correctly
+        within its scope, and also sets the distribution strategy as current for
+        the current graph & thread. Variables that need to be replicated must be
+        created within `Replicator.scope`, unless they are created inside the
+        `fn` passed to `run` which automatically wraps the entire `fn` inside
+        `Replicator.scope`. Entering a `scope` twice is safe, so in general,
+        enclosing everything that is potentially distributed inside a `scope`
+        should be safe.
+    *   `run`: This method is used to replicate the computation. It takes a
+        function `fn` which specifies the computation, and an (optional)
+        `InputIterator` `inputs` to be fed into`fn`. It calls `fn` with one
+        element from the input iterator within distribution strategy `scope`.
+        Variables and optimizers for the model can be created inside or outside
+        `fn`, as long as they're created within `scope`. `run` can be called
+        multiple times with different `fn`. In graph mode, run simply creates
+        the replicated graph whereas in eager mode, it will run the actual
+        computation. It then returns the list of outputs of `fn` from each
+        replica.
```

Comment:
A lot of these concepts seem to map to similar concepts in the Estimator world. Can you perhaps add a section explaining the difference between this API/distribution functionality and the existing Estimator API? I am concerned that we are introducing a new highish level API that looks very similar to Estimator and has many of the same usability cliffs, even though in the same breath, we are saying that Estimators are dispreferred versus Keras.

Reply:
Good point, will add a section. (I think this question still applies even with Replicator being absorbed in the core DS API). You're right that `fn` sounds similar to `model_fn` and `input_fn` sounds similar to `input_fn` in Estimator. The biggest difference is that this doesn't control the training loop whereas with Estimator, once you define, the model_fn, the training loop is run by Estimator and the user has no control over it. Examples such as the GAN and Reinforcement learning example here would be tough to do in Estimator. 
Will add a section.